### PR TITLE
Add support for Value objects being implicit dependencies

### PR DESCRIPTION
**Pull request by grossag**
As part of consolidating outside dependencies, the code I work on takes on Value objects as implicit dependencies. That allows us to take on a dependency to an entire build (e.g. our compiler) rather than 500 files from it.

This has worked fine in practice for months now, but it turns out to break when using caching, because Node.get_contents() expects all dependencies to have the "name" attribute. This change adds that attribute to the Value class and a test to confirm that Node.get_contents() works now.

## Contributor Checklist:

* [x] I have created a new test or updated the unit tests to cover the new/changed functionality.
* [x] I have updated `master/src/CHANGES.txt` directory (and read the `README.txt` in that directory)
* [x] I have updated the appropriate documentation

### CODE REVIEW

**Comment by bdbaddog** [comment]
So having a target with an implicit dependency on a Value node breaks being able to cache that target?

**Comment by grossag** [comment]
Yeah that’s right @bdbaddog

**Comment by bdbaddog** [comment]
o.k can you make a test which demonstrates that? (You've added a unit test, so this would be a system type test)

**Comment by grossag** [comment]
> o.k can you make a test which demonstrates that? (You've added a unit test, so this would be a system type test)

Ok I'll give it a try. So far it's not obvious what this would look like and where it would go, so let me know if you have any guidance you want to provide. I found test\Value.py but this doesn't seem to fit there. I found test\CacheDir folder with lots of files in it; I may see if I can add something there.

**Comment by grossag** [comment]
I'm putting this on hold for a little while so I can get a better and more representative test case. So far I haven't been able to find a good one. I'll either close or update this in the next few weeks.

**Comment by abingham** [comment]
FWIW, I can confirm that this is an issue. I have some data structures I use as dependencies when building a directory, and scons fails when looking for `name` on these values. I worked around this by just adding a `name` attribute to my nodes, but this PR looks like a more proper solution.

> I'm putting this on hold for a little while so I can get a better and more representative test case.

I can't supply the code I'm seeing this in, but I'd hate to see this PR die for lack of a "representative" test case. What exactly are we looking for? It seems to me that if `Value` nodes are legitimate dependencies of `Dir` nodes, then a test case demonstrating that this relationship works doesn't require any more justification.

File: src/engine/SCons/Node/PythonTests.py [hunk 1]
```diff
@@ -110,6 +110,17 @@ def test___init__(self):
...
+
+class ValueChildTestCase(unittest.TestCase):
+    def test___init__(self):
+        """Test support for a Value() being an implicit dependency of a Node"""
+        value = SCons.Node.Python.Value('v')
+        node = SCons.Node.Node()
+        node._func_get_contents = 2  # Pretend to be a Dir.
+        node.add_to_implicit([value])
+        contents = node.get_contents()
+        assert len(contents) > 0
```
Comment (bdbaddog) [inline src/engine/SCons/Node/PythonTests.py]: Can you add also `assert contents == '%s %s'%(value.get_csig(), value.name)`

Comment (grossag) [reply src/engine/SCons/Node/PythonTests.py]: Done, although the expected contents needs to have a \n at the end.

**Comment by grossag** [comment]
I added a system test that also reproduced the issue without the fix. I also covered a few more cases where Value nodes are dependencies either directly or as grandchildren. That will ensure that we have good coverage of this in the future.

File: src/engine/SCons/Node/Python.py [hunk 2]
```diff
@@ -96,6 +96,10 @@ def __init__(self, value, built_value=None):
         if built_value is not None:
             self.built_value = built_value
 
+        # Set a name so it can be a child of a node and not break
+        # its parent's implementation of Node.get_contents.
+        self.name = value
```
Comment (bdbaddog) [inline src/engine/SCons/Node/Python.py]: which get_contents in Node/__init__.py references a python value nodes name?

Comment (grossag) [reply src/engine/SCons/Node/Python.py]: @bdbaddog get_contents_dir uses it as a sorting key. See https://github.com/SCons/scons/blob/master/src/engine/SCons/Node/__init__.py#L216

File: src/engine/SCons/Node/Python.py [hunk 3]
```diff
@@ -177,7 +184,7 @@ def get_csig(self, calc=None):
         return contents
 
 
-def ValueWithMemo(value, built_value=None):
+def ValueWithMemo(value, built_value=None, name=None):
     global _memo_lookup_map
```
Comment (bdbaddog) [inline src/engine/SCons/Node/Python.py:188]: you need to pass name to the created Value().

Comment (grossag) [reply src/engine/SCons/Node/Python.py:188]: Ah sorry, I missed that. I have added a test that was broken with the previous code and is fixed now.

**Comment by grossag** [comment]
I fixed the issue you raised. I also updated the documentation; let me know if you think additional changes need to be made to the docs.

**Comment by bdbaddog** [comment]
Merging. There's a few corners I'll address in follow up PR.
(caching value which has name != value where there are two Value nodes with different names)

**Pull request merged**